Ensuring ineligible persons are not participants, Change #2828

stgallagher opened the pull request:
The original implementation to add the necessary behavior is [here](https://gist.github.com/greenone1975/4707723). I thought this method was doing too much, so I decomposed the behavior into as many discrete functions as I could. Tested these individually (even though they were private) and ran into some difficulty coming up with useful ways to test the composing method. 

After a discussion with Paul and Rhett, I was introduced to the [Data Context Interaction](http://rubysource.com/dci-the-evolution-of-the-object-oriented-paradigm/) paradigm, and the last commits are my attempt to glean, at least partially, and to the extent of my early understanding of the practice, some of the benefits it provides. The most important of which is the separation of logic away from where it doesn't belong, and into a more logical "process" class that's based on the particular use case I am trying to satisfy.

stgallagher committed:
broke apart #update_participant_based_on_survey into component methods, with testing #2828

stgallagher committed:
relocated ineligibility logic out of surveyor_controller, change #2828.

stgallagher committed:
added MakingIneligible to and ineligibility process, Change #2828

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
```
How about this:

``` ruby
class Participant < ActiveRecord::Base
  def ineligible!
    self.class.transaction do
      participant_person_links.destroy_all
      events.update_all(:participant_id => nil)
      # other actions as necessary
    end
  end
end

class ResponseSetObserver < ActiveRecord::Observer
  include EligibilityExamination

  def after_save(rs)
    action(rs)
  end
end

module EligibilityExamination
  def action(rs)
    person = rs.person

    if !person.eligible? && rs.survey.title =~ /PBSamplingScreen/
      person.participant.ineligible!
    end
  end
end
```

The above proposal is an improvement in a few ways:
1. it sticks to Rails conventions
2. it is clear about _what_ becomes ineligible (AFAICT, it's a Participant)
3. it avoids explicit runtime modification of classes (which can be slow, and interferes with JITs)
4. it avoids awkward names like `ineligibilifier`
5. it works for all ResponseSet modifications, not just those occurring in an HTTP request

The names of modules in my proposal do need work and the logic is suspect (in particular, eligibility examination looks wrong to me), but I don't understand why you need to introduce a new architecture.  DCI has its merits, but mixing it with existing code that doesn't follow that pattern requires that it offer a very large payoff and I am not sure what that payoff is.

rsutphin replied:
> 1. it is clear about what becomes ineligible (AFAICT, it's a Participant)

This should be made clear through naming, yes. However, it's not an argument for one architecture over another. (It's actually the person which is being marked ineligible.)

> 1. it avoids explicit runtime modification of classes (which can be slow, and interferes with JITs)

Agreed, the runtime class modification is not necessary. I'm not sure what it is trying to achieve, but certainly a DCI approach does not require it.

> 1. it works for all ResponseSet modifications, not just those occurring in an HTTP request

A DCI approach makes this logic available to all places where you need to make a Participant ineligible, not just those resulting from ResponseSet modifications.

Invoking the context from an Observer is probably a good idea, so long as it is idempotent. I can't think of a reason why it couldn't be.

> `person.participant.ineligible!`

Marking a person as ineligible affects models of several different types, including creating a `SampledPersonIneligibility` and removing the associated `Participant` and all its relationships everywhere.

One of the actions that happens when a person is recorded as ineligible is that the participant record itself is deleted. I think it's awkward to have that action be on the participant. This logic is also PBS-specific, so binding it to the generically-named method `ineligible!` on `Participant` (or `Person`) doesn't sit right.

> DCI has its merits, but mixing it with existing code that doesn't follow that pattern requires that it offer a very large payoff and I am not sure what that payoff is.

From a higher-level view, we have lots of places where there is logic which applies to several models in concert that is arbitrarily associated with a single model. It's often hard to locate the logic that applies to a particular feature. Rails does not have a convention to handle this, AFAIK. DCI provides one way, so I asked Steve to try it out here. 

I think you've identified some problems with the implementation, but I don't think that any of them apply to the general idea of using DCI.

hannahwhy replied:
> I think you've identified some problems with the implementation, but I don't think that any of them apply to the general idea of using DCI.

Yeah, they're not points against DCI.  However, if the existing architecture can support PBS ineligibility, I don't think there's a need to introduce more architecture.

I've been writing up other ideas, though, and they all tend to fall into a DCI-ish pattern, so that may be the way to go.

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
+    @ineligibilifier = Ineligibilifier.new.extend Ineligible
+  end
+
+  def make_ineligible
```
This set of actions _must_ be wrapped in a transaction.

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def disassociates_participant_from_all_events(participant)
+      Event.where( :participant_id => participant.id ).all.map { |e| e.update_attribute(:participant_id, nil) }
```
`participant.events.update_all(:participant_id => nil)` is much more efficient:
1. It does not instantiate any ActiveRecord objects
2. It compiles down to a single UPDATE SQL statement

hannahwhy replied:
This may also be clearer:

``` ruby
class Participant < ActiveRecord::Base
  has_many :events, :dependent => :nullify
end

participant.events.destroy_all   # sets all participant_id fields in associated events to NULL
```

pfriedman replied:
Wouldn't `participant.events.destroy_all   # sets all participant_id fields in associated events to NULL` also delete the the Event record? 

cf. http://apidock.com/rails/ActiveRecord/Relation/destroy_all

It looks like the original method or the suggested `participant.events.update_all(:participant_id => nil)` keeps the Event but only removed the association to the Participant.

rsutphin replied:
> Wouldn't `participant.events.destroy_all # sets all participant_id fields in associated events to NULL` also delete the the Event record? 

I believe that `:dependent => :nullify` in @yipdw's suggestion would handle this. I'd want to be sure there was a spec that verifies that the events are not deleted, though.

I'm also not sure if `:dependent => :nullify` is the least surprising default but since the relationships are so complex it is probably safer. (I.e., we're always going to need special logic to handle deletions for the specific features where removing participants or events is necessary.)

hannahwhy replied:
> Wouldn't participant.events.destroy_all # sets all participant_id fields in associated events to NULL also delete the the Event record? 

`:dependent => :nullify` alters that behavior:

``` ruby
class User < ActiveRecord::Base
  has_many :posts, :dependent => :nullify
end

class Post < ActiveRecord::Base
  belongs_to :user
end

u = User.create
5.times { u.posts.create }
u.destroy
Post.count == 5 # => true
Post.first.user_id # => nil
```

rsutphin commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
+  module Ineligible
+
+    def delete_participant_person_links(person, participant)
+      ParticipantPersonLink.where(:participant_id => participant, :person_id => person).all.each { |e| e.delete}
```
You'll want to delete all the links for that participant (not just the self link), since you're going to delete it.

rsutphin commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+end
+
+class Ineligibilifier; end
```
Why are this class & module necessary at all? Why not just implement these methods on `MakeIneligible`?

stgallagher replied:
Looking over some example DCI implementations, they usually contained 'Actor's' involved in an exchange, although in the ineligibility scenario all the potential actors (person, participant) seem too passive to be labeled as actors. Ineligibilifier was a contrived actor to fulfill this role (albeit with a stupid name). 

The module is also based on example code, where the actors are 'taught' actions specifically for a given purpose. I understood it as a way to keep actors and actions clearly defined and separate,  and also as a way to group like actions.  (It probably makes more sense when there is more than one category of actions involved in an interaction).

rsutphin commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
+    @ineligibilifier.disassociates_participant_from_all_events(@participant)
+    @ineligibilifier.disassociates_participant_from_response_set(@response_set)
+    @ineligibilifier.creates_ineligibility_record(@participant)
+  end
```
Missing here: deleting the participant. The purpose of doing all these disassociations is to enable removing the participant record, which should not exist if the person is ineligible.

rsutphin commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def creates_ineligibility_record(participant)
+      SampledPersonsIneligibility.create_from_participant!(participant)
```
Since conceptually there doesn't ever need to even be a Participant in order for a `SampledPersonsIneligibility` to be created, it might make more sense for this method to take the person record instead of a participant.

stgallagher committed:
removed extraneous actor class; pull methods out of module; wrapped #make_eligible actions in a transaction, Change #2828

stgallagher committed:
remove person as a parameter for @delete_participant_person_links, Change #2828

stgallagher committed:
a way of creating a SampledPersonsIneligibility record from person, Change #2828

stgallagher committed:
updated testing to reflect changes in implementation: events are not deleted, participant record is deleted, Change #2828

stgallagher commented:
#### Summary

Incorporated some changes based on feedback. Please review and suggest further improvements or endorse a merge.

#### Changes
- removed `Ineligibilifier` class and removed the methods from the internal module, deleting the module, calls are now to self
- added method to remove `participant` record
- refactored `Events` association removal
- deleting all the`ParticipantPersonLinks` associated with a `participant`, not just the self link
- added method on `SampledPersonIneligibility` to take `person` as an argument instead of `participant`

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  def delete_participant_person_links(participant)
+    ParticipantPersonLink.where(:participant_id => participant).all.each { |e| e.delete}
```
`ParticipantPersonLink.where(:participant_id => participant.id).delete_all`

This instantiates only `Relation` objects; it doesn't build any records.

Of course, if there are on-destroy callbacks that need to run (maybe audit trail) then you can't get away with `delete` or `delete_all`.  Instead, you will have to use `destroy` or `destroy_all`.

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,44 @@
+class MakingIneligible
```
`MakeIneligible` sounds better to me.

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,44 @@
+class MakingIneligible
+
+  def self.make_ineligible(response_set)
```
The context is established with the class name.  Since there's only one action that you can do from this context, it acts like a command, and therefore I suggest renaming this to something like `.run`.

hannahwhy commented on app/models/sampled_persons_ineligibility.rb:
```diff
@@ -43,4 +43,17 @@ def self.create_from_participant!(participant)
 
+  def self.create_from_person!(person)
+    participant = person.participant
+    code = lambda { |value| value ? 1 : 2 }
```
What do `1` and `2` mean?  Are there better names?

pfriedman replied:
If these are what I think they mean this line could be replaced with 
`code = lambda { |value| value ? NcsCode::YES : NcsCode::NO }`

hannahwhy commented on spec/context/making_ineligible_spec.rb:
```diff
@@ -0,0 +1,60 @@
+
+  it "deletes all participant_person_links for a participant" do
+    MakingIneligible.make_ineligible(@response_set)
+    ParticipantPersonLink.where(:participant_id => @response_set.participant).all.size.should be 0
```
`ParticipantPersonLink.where(:participant_id => @response_set.participant).count.should == 0`

hannahwhy commented on app/context/making_ineligible.rb:
```diff
@@ -0,0 +1,44 @@
+
+  def make_ineligible
+    ActiveRecord::Base.transaction do
+      creates_ineligibility_record(@person)
```
Since these are all single commands that don't make sense outside of the "make ineligible" use case, I suggest just rolling all five callees into this method.  Separate the lines with comments if necessary.

stgallagher replied:
With respect, I disagree in this case. My primary motivation in this branch was to take methods that were doing way to much, separate out logic, keep each step nice and functional, with one composing method that, thanks to the separation, reads like a description of the use case procedure. I don't see the advantage of cramming these otherwise unrelated lines of logic into one method. I'm not sure if the short methods need to make sense outside the context of the "make ineligible" use case. Is it not good enough that their functionality in concisely summed-up in their names?

hannahwhy replied:
I'm okay with leaving the code as-is. 

However:

>  My primary motivation in this branch was to take methods that were doing way to much, separate out logic, keep each step nice and functional, with one composing method that, thanks to the separation, reads like a description of the use case procedure. 

Moving all of that code into a single place (as this commit does) is the major contribution towards making ineligibility easier to understand.  Once that's done, though, I think

``` ruby
def make_ineligible
  transaction do
    # Build an ineligibility record for the participant's person...
    SampledPersonsIneligibility.create_from_person!(@person)

    # ...destroy all their personal relationships, ...
    ParticipantPersonLink.where(:participant_id => participant.id).delete_all

    # ... links to their events (but not the events themselves), ...
    participant.events.destroy_all

    # etc
  end
end
```

is even easier to read, because you don't have to learn any more new names.  Again, however, this is a matter of taste, and in the absence of any additional standards I'm fine with the way it is now.

> I don't see the advantage of cramming these otherwise unrelated lines of logic into one method.

That's the thing: they _are_ related.  Although they affect different models, all of that code must be run to complete ineligibility.

hannahwhy commented on spec/context/making_ineligible_spec.rb:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  it "sets the participant_id to nil for all events associated with the person" do
+    participant_id =  @response_set.participant.id
```
Correct spacing.

hannahwhy commented on spec/context/making_ineligible_spec.rb:
```diff
@@ -0,0 +1,60 @@
+    @response_set.participant.should be_nil
+  end
+
+  it "calls #create_from_particpant! from the SampledPersonsIneligibility class" do
```
What is the effect of this call?  If it doesn't have one, it may not be worth an example.

Mocks are best used when you need to test something with a well-defined interface, but getting an instance of that something to test is infeasible.  I don't think either of those are true here.

stgallagher replied:
Good Point. I was sideways testing something I could have tested directly. Fixed.

hannahwhy commented on spec/context/making_ineligible_spec.rb:
```diff
@@ -0,0 +1,60 @@
+  it "does not delete the event records" do
+    participant_id =  @response_set.participant.id
+    MakingIneligible.make_ineligible(@response_set)
+    @event1.should_not be_nil
```
This isn't what you want to test.

`@event1` will _never_ be nil, because it's never assigned to anything else.

Instead, do something like:
- `@event1.should_not be_destroyed` (not sure if that'll work)
- `Event.exists?(@event1.id).should be_true` (more direct)

hannahwhy commented on spec/context/making_ineligible_spec.rb:
```diff
@@ -0,0 +1,60 @@
+
+  it "sets the participant_id to nil for the response set" do
+    MakingIneligible.make_ineligible(@response_set)
+    @response_set.participant.should be_nil
```
ActiveRecord caches associated records until told to reload, so that may be interfering with this example.

The strongest verification is to hit the database:

``` ruby
ResponseSet.find(@response_set.id).participant_id.should be_nil
```

hannahwhy commented on app/controllers/surveyor_controller.rb:
```diff
@@ -93,11 +93,15 @@ def load_and_update_response_set_with_retries(remaining=2)
     end
   end
+
+  def person_taking_screener_ineligible?(response_set)
```
Since this method is complex enough to have specs and isn't dependent on a request, it belongs in a model or context.

stgallagher committed:
changed name of class and file to "MakeIneligible", Change #2828

stgallagher committed:
changed class method name to 'run', refactored methods, Change #2828

stgallagher committed:
remove #person_taking_screener_ineligible?(response_set) from SurveyorController, Change #2828

stgallagher committed:
added #person_taking_screener_ineligible? to Person, Change #2828

stgallagher committed:
changed magic numbers to named constants, Change #2828

stgallagher committed:
deleted unnecessary specs, Change #2828

stgallagher committed:
added specs to Person for #person_taking_screener_ineligible?, Change #2828

stgallagher committed:
refactored methods, changed filename, Change #2828

stgallagher commented:
#### Summary

Many thanks to **yipdw** for the insightful feedback. Incorporated all of the suggested changes, with exception of one noted in the previous comments, the combination of the methods in `MakeIneligible` into a single method.

_Please review changes and make further suggestions for improvements, or endorse a merging._

#### Changes
- refactored `#delete_participant_person_links` to use `#destroy_all` instead of iterating
- changed class name from `MakingIneligible` to `MakeIneligible`
- changed class method name from `make_ineligible` to `run`
- changed magic value answers in `#create_from_person!` to named constants
- refactored assertion for `ParticipantPersonLink` to use `#count` instead of `#size`
- refactored assertion for `SampledPersonsIneligibility` record creation from a call expection to a database query to confirm the record's generation
- corrected spacing in two places
- refactored `#disassociates_participant_from_all_events`: former implementation was deleting events
- refactored `#disassociates_participant_from_response_set`, it was not persisting the setting of `participant_id` on `response_set` to `nil`, changed from in-memory to database transaction assertion
- pulled `#person_taking_screener_ineligible?` out of `SurveyorController` and into `Person`

hannahwhy commented on spec/controllers/surveyor_controller_spec.rb:
```diff
@@ -154,4 +154,23 @@ def a_ui_response(hash)
+        @screener_and_ineligible_person_response_set = Factory(:response_set, :survey_id => screener_survey.id, :person => ineligible_person)
+      end
+
+      it "calls MakingIneligible if the person from the response_set is ineligible" do
```
The context's name is no longer `MakingIneligible`.

hannahwhy commented on spec/controllers/surveyor_controller_spec.rb:
```diff
@@ -154,4 +154,23 @@ def a_ui_response(hash)
+        @survey_controller = SurveyorController.new
+        screener_survey     = Factory(:survey, :title => "INS_QUE_PBSamplingScreen_INT_PBS_M3.0_V1.0")
+        ineligible_person   = Factory(:person)
+        ineligible_person.stub(:eligible?).and_return(false)
```
Shorthand:

``` ruby
ineligible_person.stub!(:eligible? => false)
```

hannahwhy commented on spec/models/person_spec.rb:
```diff
@@ -656,4 +656,36 @@
+      @non_screener_and_eligible_person_response_set   = Factory(:response_set, :survey_id => non_screener_survey.id, :person => eligible_person)
+    end
+
+    it "if survey is screener and the person is eligible, returns false" do
```
This reads better:

``` ruby
it "returns false if the survey is a screener and the person is eligible"
```

The `it` is part of the example description.

hannahwhy replied:
You can also add hierarchy for this:

``` ruby
describe 'given a screener' do
  it 'returns false if the person is eligible'
  it 'returns true if the person is ineligible'
end

describe 'given a non-screener' do
  it 'returns false'
end
```

stgallagher committed:
reworded and reorganized assertions, Change #2828

stgallagher committed:
shortened stub, fixed typo in assertion description, Change #2828

stgallagher commented:
#### Summary

Cleaned up a bit more based on feedback.

_Please review changes and make further suggestions for improvements, or endorse a merging._

#### Changes
- `MakingIneligible` typo fixed
- shortened stub and expectation statement
- reorganize and reworded assertions for `Person` spec

hannahwhy commented:
Can you combine these changes into a set of smaller changes that reflect only the final state of this pull request?

The intermediate steps are useful here, but they are not that useful for getting a look at overall project history.

hannahwhy commented:
Also, Github's merge indicator tells me that this merge will create conflicts.  Since this branch isn't that long-lived, I suggest rebasing it onto origin/master before collapsing changes.

hannahwhy commented on app/models/person.rb:
```diff
@@ -471,6 +471,11 @@ def first_child?
+
+  def person_taking_screener_ineligible?(response_set)
+    !response_set.person.eligible? && response_set.survey.title =~ /PBSamplingScreen/
+  end
```
I think this actually belongs on `ResponseSet`.  Consider this: it doesn't address the `self` person at all.

hannahwhy replied:
Maybe this:

``` ruby
class ResponseSet
  def makes_person_ineligible?
    person.eligible? && survey.title =~ /PBSamplingScreen/
  end
end

response_set.makes_person_ineligible?
```

It is a bit weird to have specific survey titles encoded in the `ResponseSet`.  You could create a new context/interaction for this.

rsutphin replied:
> Consider this: it doesn't address the `self` person at all.

This is a good principle to note, but I don't think that the alternative is better in this case. If we had a separate subclass of `ResponseSet`, putting it that subclass would make sense.

For this case, I think that neither `Person` nor `ResponseSet` is the right place. Why not have it on the context itself? This could be handled one of two ways:
- `MakeIneligible#run` could be a no-op if it wasn't appropriate to make the person ineligible
- `MakeIneligible` could have a factory method that would return an instance IFF it should apply to a a passed-in `ResponseSet`.

I prefer the latter, but either would work. Ex. for the latter:

``` ruby
class MakeIneligible
  def self.create(response_set)
    if makes_person_ineligible?(response_set)
      new([...])
    else
      nil
    end
  end
end
```

Then in the controller you'd have something like:

``` ruby
[
  MakeIneligible
].collect { |context| context.create(response_set) }.compact.each { |context| context.run }
```

Supporting a list of contexts is possible YAGNI territory, but I think that generalizing helps to think about where the seams are in this case.

pfriedman commented:
+1 on yipdw's suggestion on rebasing this branch.

rsutphin commented on app/controllers/surveyor_controller.rb:
```diff
@@ -93,11 +93,10 @@ def load_and_update_response_set_with_retries(remaining=2)
+    if response_set.person.person_taking_screener_ineligible?(response_set)
+      MakeIneligible.run(response_set)
+    else
+      response_set.participant.update_state_after_survey(response_set, psc)
```
I don't like that the remainder of this method is screened out based on whether this context is executed. It ties the controller and the context too closely together. It would be cleaner if the context were executed, then the `update_state_after_survey` call were made IFF it could still be made.

hannahwhy commented:
Why are commits from @Nataliya, @pfriedman, and myself showing up in this pull request?  They shouldn't be.

stgallagher commented:
@yipdw Yeah, I might of screwed-up the rebase...

stgallagher commented:
### Update

I'm transplanting the eligibility conditional code into the `MakeIneligibile` class, where if the conditional is true, it performs its procedure, and, if it isn't, it does nothing. 

While trying to mock out these two scenario's in testing, **yipdw** suggested making eligibility an attribute (on either `Person` or `Participant`) and building an `EligibilityObserver` class to watch for changes from the responses so it will know when to change the eligibility attribute. We also talked about possibly extracting eligibility-specific logic into its own class.

stgallagher committed:
moved conditional for ineligibility from Person to MakeIneligible #2828

stgallagher committed:
changed class name and run method #2828

changed class name and run method to reflect the
new behavior of the class. It now not only
performs the actions required for ineligibility,
it also determines eligibility within the class

stgallagher commented:
## Ready for Review

Making eligibility an attribute has been documented as a ticket for the [future](https://code.bioinformatics.northwestern.edu/issues/issues/show/3397).

rsutphin commented on app/models/participant.rb:
```diff
@@ -64,7 +64,7 @@ class << self; attr_accessor :importer_mode_on; end
   has_many :participant_consents
   has_many :participant_consent_samples
-  has_many :events
+  has_many :events, :order => 'events.event_type_code, events.event_start_date'
```
Why was this default ordering re-introduced? It was deliberately removed in 8596278e for #2820.

hannahwhy commented:
This pull request contains a lot of work from other branches, and should not be merged until that work is removed.

(Aside from currently complicating review, merging it as-is will also complicate project history.)

pfriedman commented:
@greenone1975 wrote: "Yeah, I might of screwed-up the rebase.."

@yipdw wrote: "This pull request contains a lot of work from other branches, and should not be merged until that work is removed.

(Aside from currently complicating review, merging it as-is will also complicate project history.)"

Any thoughts about how to proceed with this branch?

rsutphin commented:
@greenone1975 and I successfully rebased onto master. It looks clean now.

stgallagher commented:
Rhett re-rebased and all is well now.
